### ISSUE TITLE
C backend on behavior tests emits C code with compile errors when using -std=c99 -Werror -pedantic

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#11651]
### Zig Version

0.10.0-dev.2211+f32928c50

### Steps to Reproduce

```
stage2/bin/zig test ../test/behavior.zig -I../test -ofmt=c -femit-bin=test.c
stage2/bin/zig cc -c test.c -Werror -pedantic -std=c99
```

### Expected Behavior

No output, exit code 0.

### Actual Behavior

```
test.c:937:17: error: redefinition of typedef 'zig_A_u8_1' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_1[1];
                ^
test.c:935:17: note: previous definition is here
typedef uint8_t zig_A_u8_1[1];
                ^
test.c:993:17: error: redefinition of typedef 'zig_A_u8_4' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_4[4];
                ^
test.c:934:17: note: previous definition is here
typedef uint8_t zig_A_u8_4[4];
                ^
test.c:1019:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1052:17: error: redefinition of typedef 'zig_A_u8_3' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_3[3];
                ^
test.c:959:17: note: previous definition is here
typedef uint8_t zig_A_u8_3[3];
                ^
test.c:1106:17: error: redefinition of typedef 'zig_A_u8_5' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_5[5];
                ^
test.c:936:17: note: previous definition is here
typedef uint8_t zig_A_u8_5[5];
                ^
test.c:1198:17: error: redefinition of typedef 'zig_A_u8_10' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_10[10];
                ^
test.c:989:17: note: previous definition is here
typedef uint8_t zig_A_u8_10[10];
                ^
test.c:1281:17: error: redefinition of typedef 'zig_A_u8_8' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_8[8];
                ^
test.c:925:17: note: previous definition is here
typedef uint8_t zig_A_u8_8[8];
                ^
test.c:1317:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1332:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1352:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1367:17: error: redefinition of typedef 'zig_A_i32_5' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1114:17: note: previous definition is here
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1370:17: error: redefinition of typedef 'zig_A_i32_5' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1367:17: note: previous definition is here
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1379:17: error: redefinition of typedef 'zig_A_u8_6' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_6[6];
                ^
test.c:1115:17: note: previous definition is here
typedef uint8_t zig_A_u8_6[6];
                ^
test.c:1458:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1517:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1687:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
test.c:1720:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
test.c:1735:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
test.c:1751:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

**Comment by matu3ba** [comment]
In root folder:
`./stage2/bin/zig test test/behavior.zig -Itest -ofmt=c -femit-bin=test.c`
`./stage2/bin/zig cc -c test.c -Werror -pedantic -std=c99`

zig_A_u8_3 & co are emitted multiple times, which sound like the more simple parts.

`./stage2/bin/zig cc -c test.c -Werror -Wall -std=c99` also shows up unused variables:
```txt
test.c:2766:18: error: incompatible redeclaration of library function 'printf' [-Werror,-Wincompatible-library-redeclaration]
ZIG_EXTERN_C int printf(uint8_t const * a0, ...);
                 ^
test.c:2766:18: note: 'printf' is a builtin with type 'int (const char *, ...)'
test.c:2781:46: error: declaration of built-in function 'fopen' requires inclusion of the header <stdio.h> [-Werror,-Wbuiltin-requires-header]
ZIG_EXTERN_C zig_S_behavior_bugs_4328_FILE * fopen(uint8_t const * a0, uint8_t const * a1);
                                             ^
test.c:6310:18: error: unused variable 't1' [-Werror,-Wunused-variable]
 uint8_t * const t1 = (uint8_t * )&t0;
                 ^
test.c:6712:24: error: unused variable 't1' [-Werror,-Wunused-variable]
 static uint32_t const t1[5] = {};
                       ^
test.c:7121:23: error: unused variable 't19' [-Werror,-Wunused-variable]
 static uint8_t const t19[4] = {};
                      ^
test.c:7480:20: error: unused variable 't1' [-Werror,-Wunused-variable]
 static bool const t1[11] = {};
                   ^
test.c:11007:23: error: unused variable 't0' [-Werror,-Wunused-variable]
 static uint8_t const t0[100] = {};
                      ^
test.c:11218:23: error: unused variable 't3' [-Werror,-Wunused-variable]
 static uint8_t const t3[20] = {};
                      ^
test.c:11217:23: error: unused variable 't1' [-Werror,-Wunused-variable]
 static uint8_t const t1[20] = {};
                      ^
test.c:12311:19: error: unused variable 't6' [-Werror,-Wunused-variable]
   uint32_t const t6 = (t4).payload;
                  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

The C file is quite big (66k LOC) and the naming is very convoluted.
Take as example `zig_S_behavior_struct_test_call_20method_20with_20mutable_20reference_20to_20struct_20with_20no_20fields__struct_6495`.
I dont find anything on grepping for 6495, ie with `rg 6495` inside the test folder.

Do you have an opinion on a sane way to reference the files?
I would imagine that detecting source of regressions in CI will be painful with this file size.

**Comment by matu3ba** [comment]
Bisecting gives me as minimal example with redefinition error
````zig
// tmp.zig
const std = @import("std");
const mem = std.mem;
const expect = std.testing.expect;

test "single-item pointer to array indexing and slicing" {
    var array = "aaaa".*;
    array[1] = 'z';
    array[2..3][0] = 'y';
    try expect(mem.eql(u8, "azya", &array));
}

test "implicit cast zero sized array ptr to slice" {
    var b = "".*;
    const c: []const u8 = &b;
    try expect(c.len == 0);
}
```
with commands
`./stage2/bin/zig test tmp.zig -Itest -ofmt=c -femit-bin=test.c`
`./stage2/bin/zig cc -c test.c -Werror -std=c99`
output
```txt
test.c:1611:17: error: redefinition of typedef 'zig_A_u8_1' is
 a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_1[1];
                ^
test.c:1610:17: note: previous definition is here
typedef uint8_t zig_A_u8_1[1];
                ^
test.c:1852:54: error: integer literal is too large to be repr
esented in a signed integer type, interpreting as unsigned [-W
error,-Wimplicitly-unsigned-literal]
   t23.field_1 = zig_addo_u64(t22, 1u, &t23.field_0, 184467440
73709551615);
                                                     ^
test.c:1931:60: error: integer literal is too large to be repr
esented in a signed integer type, interpreting as unsigned [-W
error,-Wimplicitly-unsigned-literal]
         t20.field_1 = zig_addo_u64(t19, 1u, &t20.field_0, 184
46744073709551615);
                                                           ^
test.c:1949:60: error: integer literal is too large to be repr
esented in a signed integer type, interpreting as unsigned [-W
error,-Wimplicitly-unsigned-literal]
         t25.field_1 = zig_addo_u64(t24, 1u, &t25.field_0, 184
46744073709551615);
                                                           ^
test.c:1978:54: error: integer literal is too large to be repr
esented in a signed integer type, interpreting as unsigned [-W
error,-Wimplicitly-unsigned-literal]
   t30.field_1 = zig_addo_u64(t29, 1u, &t30.field_0, 184467440
73709551615);
                                                     ^
5 errors generated.
```

**Comment by leroycep** [comment]
We are getting duplicate typedefs because both `.array_u8_sentinel_0` and `.array_u8` generate the same typedef:

```
array_u8_sentinel_0     "typedef uint8_t zig_A_u8_5[5];\n"
           array_u8     "typedef uint8_t zig_A_u8_1[1];\n"
```

To zig the types are distinct, but for the C backend they are lowered to the same thing, so we end up with two identical typedefs.

Thanks to matu3ba for the minimal example! Made figuring this out easier.

You can see the full debug output here: https://gist.github.com/leroycep/7558f9b35083887215fcb49256dc23e2

**Comment by leroycep** [comment]
How should the overlapping array types be fixed? I have a few ideas:

- Append the sentinel values to the type name. That is what I did in #11693, though it is too naive at the moment. It works in the case of the array having a null value (`0`), but not more complex sentinels.
- Append the sentinel value hash to the type name. Solves the problem, but the name would be unreadable.
- Deduplicate types in the C output. Feels icky to me, but :shrug:

**Comment by matu3ba** [comment]
@leroycep Are you on discord or IRC? On discord is a channel dedicated to the C backend.

Sorry, if this reads more like a little rant. Reading the C code just makes me sad.

I think that having symbol names more easily associative to the Zig file at least at function level would help a lot to figure out what would be the best options. We can utilize the new test runner ones we have a naming schema decided.

For example this is horrible and  those long and non-searchable names, I feel like fixes are disimprovements/(in german Verschlimmbesserungen):
`zig_S_behavior_struct_test_call_20method_20with_20mutable_20reference_20to_20struct_20with_20no_20fields__struct_6495`

1. It would be nice, if the C namespace references (struct, union) be immediately clearly visible and the same for the function names in the original Zig file.
2. I dont understand the numbering schema of numbers and characters. When is used what?
3. comptime functions may have multiple instances or return types, but we mostly care how they associate to the Zig code
4. common stuff like pointer slice should be lowered specially to remove struct redundancy (ptr len) and related ptr types.
5. The most annoying part: ` /* file:2:25 */` doesnt correspond to the source file name and its very sad to not being able to jump between Zig and C file. There should be at least for every top level decl a corresponding link to jump in between both, for example behind a debug flag.

Now question to the reader, if the reader made it until here:
What is `array_u8_sentinel_0     "typedef uint8_t zig_A_u8_5[5];\n"` corresponding to in the Zig code ? There is no symbol `A`.

**Comment by matu3ba** [comment]
I'll first fix #11849 to make reasoning about what the generated code associates to sane before fixing this piece by piece.

I would expect that we can use with properly working links reduction methods on the Zig files.
